torch/csrc/autograd/variable.h

```diff
@@ -1,150 +1,255 @@
 
-  // Implicitly casts a Tensor to a Variable. This should only be called on
-  // Tensors which you know are actually Variables.
-  /*implicit*/ Variable(Tensor const & rhs) : Tensor(rhs) {}
-  /*implicit*/ Variable(Tensor && rhs) noexcept : Tensor(std::move(rhs)) {}
+  /// Get the gradient accumulator of the `Variable` if it has one, or else
+  /// create one on the fly and return it.
+  std::shared_ptr<Function> grad_accumulator() const;
 
+  /// Set the gradient edge -- i.e. `grad_fn` and `input_nr` -- of the
+  /// `Variable`.
+  /// NOTE: This will always set the `grad_fn`, even if this is a leaf
+  /// variable, and never the `grad_accumulator`. For the latter, use
+  /// `set_grad_accumulator`. This allows late construction of an interior
+  /// `Variable`.
+  void set_gradient_edge(Edge&& edge);
+
+  /// Return the "canonical" gradient edge of this `Variable`, i.e. either the
+  /// gradient function if this is an interior `Variable`, or the gradient
+  /// accumulator otherwise. If the `Variable` is interior, the returned `Edge`
+  /// will store the input index of the `Function` to which this variable is
+  /// connected in its `input_nr` field. For leaves, the `input_nr` is always
+  /// zero. Note that `set_gradient_edge` and `gradient_edge` are not
+  /// symmetric. You must use `set_gradient_edge` to set the `grad_fn` and
+  /// `set_grad_accumulator` to set the accumulator.
   Edge gradient_edge() const {
     // If grad_fn is null (as is the case for a leaf node), we instead
-    // interpret the gradient function to be a grad accumulator,
-    // which will accumulate its inputs into the grad property of the
-    // variable. These nodes get suppressed in some situations,
-    // see "suppress grad accumulation" below. Note that only variables which
-    // have `requires_grad = True` can have grad accumulators.
+    // interpret the gradient function to be a gradient accumulator, which will
+    // accumulate its inputs into the grad property of the variable. These
+    // nodes get suppressed in some situations, see "suppress gradient
+    // accumulation" below. Note that only variables which have `requires_grad =
+    // True` can have gradient accumulators.
     if (const auto& gradient = grad_fn()) {
       return Edge(gradient, output_nr());
     } else {
       return Edge(grad_accumulator(), 0);
     }
   }
 
-  inline VariableImpl* get() const;
+  /// Return the input index of the gradient `Function` to which this `Variable`
+  /// is connected.
+  int output_nr() const noexcept;
 
-  inline const Tensor & data() const;
-  inline       Tensor & data();
+  void set_requires_grad(bool requires_grad);
+  bool requires_grad() const noexcept;
 
-  inline Tensor opt_data() const;
+  PyObject* pyobj() const noexcept;
+  void set_pyobj(PyObject* pyobj);
 
-  inline const Variable & grad() const;
-  inline       Variable & grad();
+  /// Set the type of the underlying `Tensor`.
+  void set_type(at::Type*);
```

Comment:
Uuuum, what? How can you set a type of the tensor without touching the tensor? Is it a type cast or what?

Reply:
This replaces [this](https://github.com/pytorch/pytorch/blob/a8bda67ff1f4229077342c429f94673b919b9e8f/torch/csrc/autograd/python_variable.cpp#L264) rather gross hack, which should definitely go away (I didn't even know it was possible at all to set private members in C++). Is there a better way?

Reply:
I agree this an awful piece of code, but making a hack part of a public API kind of makes it seem nice while what you're doing is terrible. I don't think there's a clean way to solve this, so I'd rather have the hack in the single place where it has to be, and don't imply that it's safe to do this in general.